src/QsCompiler/CompilationManager/TypeChecking.cs

```diff
@@ -2015,6 +2015,12 @@ QsCallable GetCallable((QsQualifiedName, ImmutableArray<QsSpecialization>) specI
                 }
                 compilation.UpdateCallables(callables);
                 compilation.UpdateTypes(types);
+                foreach (var (parent, diag) in compilation.VerifyCycles(callables))
+                {
+                    var info = callableDeclarations[parent];
+                    var offset = info.Position is DeclarationHeader.Offset.Defined pos ? pos.Item : null;
+                    diagnostics.Add(Diagnostics.Generate(info.SourceFile.Value, diag, offset));
```

Comment:
I am curious why the call graph can't make diagnostics with absolute ranges? In my experience, if you have a `QsCallable`, you should have enough information to get absolute positions.

Reply:
The QsCallables here are not defined with locations. The way I have added the offset here is the way it is done further up in the same method. I did it this way to maintain consistency with the other logic in this method. The reason why the callables being operated on don't have locations likely has something to do with them coming from a fragment tree rather than a proper syntax tree, but the details on that are a bit out of scope for what I needed to accomplish here.

Reply:
Sorry, I didn't see that the callables above are created with a null location. Is there any way to avoid the duplicated code here and in GetCallable above? Maybe GetCallable can use the offset it finds (plus the range, which I think is also in the DeclarationHeader?) to fill in the callable location?

Reply:
I can add the location information to the callables and the specializations on the callables, but I don't really know why they were null before. It seems like it was an intentional choice. But by doing that I do get the correct position info for the diagnostic.

Reply:
We decided offline that this change would require Compiler Diagnostics to support Source File information.